OCaml 5.1.x: out of range branch on s390x

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

The combination of OCaml 5.1.0 or 5.1.1, coq 8.18.0, and flocq 4.1.3 reliably segfaults for me when trying to build for Fedora Rawhide on s390x.  The segfault happens at the end of camlMicromega_plugin__Coq_micromega.xwitness_list_25687.  There is a brasl instruction that jumps to an invalid address, leading to the segfault.  On an x86_64 machine, the equivalent call instruction jumps to caml_call_realloc_stack.  With GDB, I see that the brasl instruction jumps to 0x208025ca320, but caml_call_realloc_stack is actually at 0x2aa025ca320, a difference of 0xa200000000.  The brasl instruction in question is emitted at asmcomp/s390x/emit.mlp line 754 in the OCaml 5.1.1 release.  My understanding is that the brasl instruction can only handle 32-bit offsets, and this particular jump is more than a 32-bit offset away, so it appears that the upper bits are silently dropped.

For what it's worth, I do not have any trouble building coq 8.18.0 itself.  All of its tests pass.  I also don't see this happen with coq 8.17.1 + flocq 4.1.1.

Note that the Fedora build of OCaml 5.1.1 contains the changes in https://github.com/ocaml/ocaml/pull/11594, since the test suite was failing without it.

--- a/Changes
+++ b/Changes
@@ -707,6 +707,9 @@ Working version
   (Olivier Nicole, review by Stefan Muenzel, Miod Vallat, Guillaume
    Munch-Maccagnoni, Gabriel Scherer and Xavier Leroy)
 
+- #12831: Fix call to caml_call_realloc_stack for s390x in PIC mode
+  (Vincent Laviron, report by Jerry James, review by Miod Vallat)
+
 OCaml 5.1.1
 -----------
 

--- a/asmcomp/s390x/emit.mlp
+++ b/asmcomp/s390x/emit.mlp
@@ -752,7 +752,7 @@ let fundecl fundecl =
       `	lay	%r15, -8(%r15)\n`;
       `	stg	%r14, 0(%r15)\n`;
       `	lgfi	%r12, {emit_int s}\n`;
-      `	brasl	%r14, {emit_symbol "caml_call_realloc_stack"}\n`;
+      emit_call "caml_call_realloc_stack";
       `	lg	%r14, 0(%r15)\n`;
       `	la	%r15, 8(%r15)\n`;
       `	brcl	15, {emit_label ret}\n`

